internal/blockchain/cardano/cardano.go

```diff
@@ -0,0 +1,653 @@
+)
+
+type Cardano struct {
+	ctx                context.Context
+	cancelCtx          context.CancelFunc
+	pluginTopic        string
+	metrics            metrics.Manager
+	capabilities       *blockchain.Capabilities
+	callbacks          common.BlockchainCallbacks
+	client             *resty.Client
+	streams            *streamManager
+	streamID           string
+	closed             chan struct{}
+	wsconn             wsclient.WSClient
+	cardanoconnectConf config.Section
+	subs               common.FireflySubscriptions
+}
+
+type ffiMethodAndErrors struct {
+	method *fftypes.FFIMethod
+	errors []*fftypes.FFIError
+}
+
+type cardanoWSCommandPayload struct {
+	Type        string `json:"type"`
+	Topic       string `json:"topic,omitempty"`
+	BatchNumber int64  `json:"batchNumber,omitempty"`
+	Message     string `json:"message,omitempty"`
+}
+
+type Location struct {
+	Address string `json:"address"`
+}
+
+var addressVerify = regexp.MustCompile("^addr1|^addr_test1|^stake1|^stake_test1")
+
+func (c *Cardano) Name() string {
+	return "cardano"
+}
+
+func (c *Cardano) VerifierType() core.VerifierType {
+	return core.VerifierTypeCardanoAddress
+}
+
+func (c *Cardano) Init(ctx context.Context, cancelCtx context.CancelFunc, conf config.Section, metrics metrics.Manager, cacheManager cache.Manager) (err error) {
+	c.InitConfig(conf)
+	cardanoconnectConf := c.cardanoconnectConf
+
+	c.ctx = log.WithLogField(ctx, "proto", "cardano")
+	c.cancelCtx = cancelCtx
+	c.metrics = metrics
+	c.capabilities = &blockchain.Capabilities{}
+	c.callbacks = common.NewBlockchainCallbacks()
+	c.subs = common.NewFireflySubscriptions()
+
+	if cardanoconnectConf.GetString(ffresty.HTTPConfigURL) == "" {
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "url", cardanoconnectConf)
+	}
+
+	wsConfig, err := wsclient.GenerateConfig(ctx, cardanoconnectConf)
+	if err == nil {
+		c.client, err = ffresty.New(c.ctx, cardanoconnectConf)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	c.pluginTopic = cardanoconnectConf.GetString(CardanoconnectConfigTopic)
+	if c.pluginTopic == "" {
+		return i18n.NewError(ctx, coremsgs.MsgMissingPluginConfig, "topic", "blockchain.cardano.cardanoconnect")
+	}
+
+	if wsConfig.WSKeyPath == "" {
+		wsConfig.WSKeyPath = "/ws"
+	}
+	c.wsconn, err = wsclient.New(ctx, wsConfig, nil, c.afterConnect)
+	if err != nil {
+		return err
+	}
+
+	c.streams = newStreamManager(c.client, c.cardanoconnectConf.GetUint(CardanoconnectConfigBatchSize), uint(c.cardanoconnectConf.GetDuration(CardanoconnectConfigBatchTimeout).Milliseconds()))
+
+	stream, err := c.streams.ensureEventStream(c.ctx, c.pluginTopic)
+	if err != nil {
+		return err
+	}
+
+	log.L(c.ctx).Infof("Event stream: %s (topic=%s)", stream.ID, c.pluginTopic)
+	c.streamID = stream.ID
+
+	c.closed = make(chan struct{})
+	go c.eventLoop()
+
+	return c.wsconn.Connect()
+}
+
+func (c *Cardano) StartNamespace(ctx context.Context, namespace string) (err error) {
+	// TODO: Implement
+	return nil
+}
```

Comment:
In FireFly v1.3.0 we move to each namespace having it's own event stream in each plugin, instead of a shared event stream for all namespace for better data isolation. 

Want to understand if this just a future thing or there is a limitation

Reply:
I was following the example of the Tezos plugin because it was simpler. I guess it was simpler because it just had one hard-coded stream. I didn't actually realize, looking at EVM, that there was supposed to be a 1:1 mapping between firefly namespaces and connector eventstreams.

I can make that change, it shouldn't affect the connector design at all.

Reply:
done